File: dsa/educosys/decodeWays.cpp
#include <iostream>
using namespace std;

int helper(string s, int ind)
{
    if (ind == s.size())
    {
        return 1;
    }

    if (ind > s.size())
        return 0;

    if (s[ind] == '0') return 0;
    // consider single digit
    int ways = helper(s, ind + 1);

    // consider double digit
    if (s[ind] == '1' || (s[ind] == '2' && s[ind + 1] >= '0' && s[ind + 1] <= '6'))
        ways += helper(s, ind + 2);

    return ways;
}

int numDecodings(string s)
{
    return helper(s, 0);
}

int main(void)
{
    string s = "26";
    cout << numDecodings(s) << endl;
    return 0;
}

File: dsa/educosys/DecodeWays.java
package educosys;

public class DecodeWays
{

    public static int helper(String s, int ind, int n)
    {
        if (ind >= n - 1) return 1; 

        int ways = 0;    

        // single digit    
        if (s.charAt(ind) != '0') 
        	ways += helper(s, ind + 1, n);    

        // double digit    
        if ((s.charAt(ind) == '1' || (s.charAt(ind) == '2' && s.charAt(ind + 1) <= '6')))     
            ways += helper(s, ind + 2, n);

        return ways;
    }
    public static int numDecodings(String s) {    
        if (s.charAt(0) == '0') return 0;    
        int n = s.length();
        return helper(s, 0, n);
    }

    public static void main(String[] args)
    {
        System.out.println(numDecodings("10"));
    }

}
